### PR TITLE
[oneDNN] Further ops refactoring of oneDNN cache access

### PR DESCRIPTION
**Pull request by jczaja**
### PR types
Performance optimization

### PR changes

OPs

### Describe

1. Refactoring of BatchNorm and Activation to comply with Acquire API.
2. isCached was removed and  isCachedNonBlocking was renamed to isCached. Same for Acquire functions.
3. All ops are now to use simplified isCached().

### CODE REVIEW

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 1]
```diff
@@ -178,7 +169,15 @@ class MKLDNNHandlerT {
     bwd_pd_ = std::static_pointer_cast<typename TBackward::primitive_desc>(
```
Comment (lidanqing-vv) [inline paddle/fluid/platform/mkldnn_reuse.h:169]: Hi I saw you name key_fpd, Maube you could name here backward pd key as key_bpd ?

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 2]
```diff
@@ -178,7 +169,15 @@ class MKLDNNHandlerT {
...
 
-    return (bwd_pd_ != nullptr);
+    if (bwd_pd_ == nullptr) {
+      return false;
+    } else {
+      // When BWD is cached then still we need to Get FWD PD
+      const std::string key_fpd = key_ + "@fwd_pd";
+      fwd_pd_ = std::static_pointer_cast<typename TForward::primitive_desc>(
+          dev_ctx_.GetBlob(key_fpd));
+      return true;
```
Comment (lidanqing-vv) [inline paddle/fluid/platform/mkldnn_reuse.h:182]: fwd_pd_ get value by dev_ctx_.GetBlob, then could it be null_ptr? If it is null will it still return true or something like "return bwd_pd_ == nullptr ? false : true"  ?

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 3]
```diff
@@ -187,29 +186,6 @@ class MKLDNNHandlerT {
...
-                                         std::forward<Args>(args)...);
-        dev_ctx_.SetBlob(key_pd, fwd_pd_);
-      }
-    }
-  }
-
-  template <typename Arg, typename... Args>
-  void AcquireForwardPrimitiveDescriptorNonBlocking(Arg&& first_arg,
-                                                    Args&&... args) {
     // This is used when we can recreate FWD PD in BWD so
```
Comment (lidanqing-vv) [inline paddle/fluid/platform/mkldnn_reuse.h:192]: These two lines comments are describing `void AcquireForwardPrimitiveDescriptor(Arg&& first_arg, Args&&... args)` ?

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 4]
```diff
@@ -242,31 +218,10 @@ class MKLDNNHandlerT {
...
-          bwd_desc, engine_, *fwd_pd_);
-      dev_ctx_.SetBlob(key_pd, bwd_pd_);
-    }
-  }
-
-  template <typename... Args>
-  void AcquireBackwardPrimitiveDescriptorNonBlocking(Args&&... args) {
     // fwd_pd_ is set during grad by calling
-    // AcquireForwardPrimitiveDescriptorNonBlocking
+    // AcquireForwardPrimitiveDescriptor
```
Comment (lidanqing-vv) [inline paddle/fluid/platform/mkldnn_reuse.h:227]: Does these comments "// fwd_pd_ is set during grad by calling  // AcquireForwardPrimitiveDescriptor" still suit here? Here it looks only bwd_pd_ is created and cached.

**Review by arogowie-intel — DISMISSED** [review]

**Review by arlesniak — DISMISSED** [review]
LGTM

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 2 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h:182]: if bwd_pd_ is valid so then should be fwd_pd_ . But it is a valid point to add some PADDLE_ENFORCE making sure that fwd_pd_ is also valid .

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h:192]: It is still a relevant comment to this function

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h:169]: Normally key_pd is used everywhere it just at that particular moment I needed two keys so I added key_fpd.

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 4 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h:227]: AcquireBackwardPrimitiveDescriptor is creating BWD_PD only that is correct , but when implementing Handler for grad(backward) you also call AcquireForwardPrimitiveDescriptor which will create FWD_PD. So in grad handler we call AcquireForwardPrimitiveDescriptor and AcquireBackwardPrimitiveDescriptor

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 2 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h:182]: @lidanqing-intel I added PADDLE_ENFORCE to make sure fwd_pd_ is set when bwd_pd_ is cached. Please sontinue your review

**Review by lidanqing-vv — DISMISSED** [review]
LGTM. Thanks a lot for explanation !

**Comment by jczaja** [comment]
@juncaipeng Hi, There is a problem PR-CI-Windows build. I had it rerunning three times and still it failes due to not related to PR changes errors:

2021-06-12 22:50:51   failed with:
**2021-06-12 22:50:51    ninja: error: failed recompaction: Permission denied**
2021-06-12 22:50:51 CMake Error at C:/Program Files/CMake/share/cmake-3.17/Modules/CheckIncludeFileCXX.cmake:92 (try_compile):
2021-06-12 22:50:51   Failed to generate test project build system.

Please help

**Review by arlesniak — DISMISSED** [review]
LGTM

**Review by jakpiase — DISMISSED** [review]
LGTM

**Comment by lidanqing-vv** [comment]
@jczaja Wangzhen said it was Windows machine issue last week and now it's recovered.

But me and wanghzen both can not rerun it because CI says: "**This commit exceeds rerun limitation.**" 

Please change this PR a bit and submit a new commit so CI could restart.

**Review by lidanqing-vv — APPROVED** [review]
LGTM

**Comment by jczaja** [comment]
@juncaipeng This CheckPRTemplate is being running for many hours .  Could you please advice how to make it finished?

**Comment by juncaipeng** [comment]
> @juncaipeng This CheckPRTemplate is being running for many hours . Could you please advice how to make it finished?

In this case, it is necessary to modify the describe of the PR, so the CheckPRTemplate will rerun.

**Pull request merged**